RHIDP-12524: Platform engineer define, name, and view aggregated KPIs for homepage (#2183)
redhat-developer/red-hat-developers-documentation-rhdh: modules/observability_evaluate-project-health-using-scorecards/proc-configure-aggregated-kpis-for-the-scorecard.adoc

Reviewer: Want to clarify: am I right in understanding that this chapter explains how to configure an aggregated scorecard? 

If so, I would recommend explaining how customers can configure:
- An aggregation card with a `default configuration`:
   - In addition to using `app-config.yaml`, the customer must provide a configuration to the **dynamic plugin configuration file**, as shown in the [linked example](https://redhat-developer.github.io/red-hat-developers-documentation-rhdh/pr-2183/observability_evaluate-project-health-using-scorecards/#configure-aggregated-scorecard-kpis-to-monitor-the-health-of-multiple-applications_monitor-collective-health-across-components:~:text=h%3A%206%20%7D-,%2D%20mountPoint%3A%20home.page/cards%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20importName%3A%20ScorecardHomepageCard%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20config%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20props,h%3A%206%20%7D%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20xxs%3A%20%7B%20w%3A%2012%2C%20h%3A%206%20%7D,-%2D%20package%3A%20./dynamic). If you look at this config, the `aggregationId` should be used instead of `metricId`, which contains the metric value. For example:  
``` yaml
    - mountPoint: home.page/cards
      importName: ScorecardHomepageCard
      config:
        props:
          aggregationId: "github.open_prs"
        layouts:
          xl: { w: 3, h: 6, x: 3 }
          lg: { w: 4, h: 6, x: 4 }
          md: { w: 6, h: 6, x: 6 }
          sm: { w: 12, h: 6 }
          xs: { w: 12, h: 6 }
          xxs: { w: 12, h: 6 }
```
   - In case of using metricId value at the `aggregationId` attribute will be used default config. No extra configuration required only on the **dynamic plugin configuration file** configuration in `app-config.yaml` is not needed. For more information use [this doc](https://github.com/redhat-developer/rhdh-plugins/blob/main/workspaces/scorecard/plugins/scorecard-backend/docs/aggregation.md#:~:text=If%20there%20is,and%20app%2Dconfig.)
- An aggregation card with the `statusGrouped` type:
   - To use this type customer should configure `app-config.yaml` (`scorecard.aggregationKPIs`) and **dynamic plugin configuration file** (props.aggregationId). For more context I recommend you to read the whole information about [Aggregation types in doc](https://github.com/redhat-developer/rhdh-plugins/blob/main/workspaces/scorecard/plugins/scorecard-backend/docs/aggregation.md#aggregation-types)
- An aggregation card with the `average` type.
   - The same as for `statusGrouped` type

Author: I have updated the PR to reflect your suggestions.

Reviewer: Thank you, I like the chapter descriptions!

:_mod-docs-content-type: PROCEDURE

[id="configure-aggregated-kpis-for-the-scorecard_{context}"]
= Configure aggregated KPIs for the scorecard

[role="_abstract"]
Define aggregated Key Performance Indicators (KPIs) in your configuration to provide a consolidated view of team or group health. Aggregating KPIs allows you to summarize technical metrics into high-level insights for management and stakeholders.

.Prerequisites
* The scorecard plugin is installed and configured in your {product} instance.
* You have identified the `metricId` you wish to aggregate.

.Procedure

. Access your `app-config.yaml` file.

. Navigate to the `scorecard` section and add an `aggregationKPIs` block.

. Update the configuration to create a homepage card that summarizes a specific metric for a team portfolio. The following example adds a Jira open issues card that groups counts by threshold status for entities the signed-in user owns:
+
[source,yaml]
----
scorecard:

>   aggregations:

Reviewer 2: ```suggestion
  aggregationKPIs:
```

Looks like this one slipped through

Author: thanks!

    openIssuesKpi:
      title: 'Jira open issues KPI'
      description: 'Open issues across entities you own, grouped by status.'
      type: statusGrouped
      metricId: jira.open_issues
----
+
where:
+
`scorecard.aggregationKPIs`:: Map of KPI keys to KPI definitions. Each key is an aggregation ID. For example `openIssuesKpi`.
`title`:: Title shown for the KPI.
`description`:: Short description of the KPI.
`type`:: Aggregation strategy. Use `statusGrouped` for counts per threshold status, or `average` for a weighted portfolio score.
`metricId`:: Metric provider ID. For example `jira.open_issues`, `github.open_prs`.

. Save the configuration and restart your {product} instance.

.Verification

. Navigate to the {product-short} homepage.

> . Verify that the new aggregated card is displayed with the defined name and calculated value.

Reviewer: I like this information. I would say that sub-chapter about "An aggregation card with the statusGrouped type" is ready. Only the description about `type` may be updated to describing only `statusGrouped` as we may want to have different sub-chapter for each types.

I would recommend use similar information for `average` type, but add information that:
- `options.statusScores`
  - `required` for average type 
  - map threshold rule keys (status keys for example **success**, **warning**, **error**) to numeric weights. 
  - Must be non-empty. Missing options / empty statusScores fails startup.
- `options.threshold`:
   - `Optional` same-shape number thresholds as metrics, evaluated against averageScore on 0–100 (one decimal).
   - If omitted, AverageAggregationStrategy uses built-in defaults: <30 error, 30–79 warning, ≥80 success (documented in aggregation.md and thresholds.md).
   - If you do set thresholds, they must obey full real-line coverage (no gaps), same as number metric thresholds.

Author: added topics to PR to reflect this

:_mod-docs-content-type: PROCEDURE

[id="configure-aggregated-kpis-for-the-scorecard_{context}"]
= Configure aggregated KPIs for the scorecard

[role="_abstract"]
Define aggregated Key Performance Indicators (KPIs) in your configuration to provide a consolidated view of team or group health. Aggregating KPIs allows you to summarize technical metrics into high-level insights for management and stakeholders.

.Prerequisites

> * The scorecard plugin is installed and configured in your Red Hat Developer Hub instance.

Reviewer 2: ```suggestion
* The scorecard plugin is installed and configured in your {product} instance.
```

I believe that the preference is for `{product}` over "Red Hat Developer Hub"

Author: Thanks, I completely missed this one! Done.

> * You have identified the `metricId` and the `owner` group from the Software Catalog you wish to aggregate.

Reviewer 2: ```suggestion
* You have identified the `metricId` you wish to aggregate.
```

Users are unable to specify the owners for the aggregation logic. It is based the ownership references of the viewer

Author: done

.Procedure

. Access your `app-config.yaml` file.

> . Navigate to the `scorecard` section and add an `aggregations` block.

Reviewer 2: ```suggestion
. Navigate to the `scorecard` section and add an `aggregationKPIs` block.
```

Update to match the actual config value

Author: done

> . Update the configuration to create a homepage card that summarizes a specific metric for a team portfolio. The following example calculates the total bug count for all components owned by Team Alpha:

Reviewer 2: ```suggestion
. Update the configuration to create a homepage card that summarizes a specific metric for a team portfolio. The following example adds a Jira open issues card that groups counts by threshold status for entities the signed-in user owns:
```

Same as above, it is based on the viewer

Author: done

+
[source,yaml]
----
scorecard:
  aggregations:
    - id: team-alpha-total-bugs
      name: "Total Open Bugs (Team Alpha)"
      description: "Aggregated sum of Jira bugs across all Team Alpha components"
      type: sum
      metricId: jira.open_bugs
      selector:

>         owner: group:default/team-alpha

Reviewer 2: ```suggestion
scorecard:
  aggregationKPIs:
    openIssuesKpi:
      title: 'Jira open issues KPI'
      description: 'Open issues across entities you own, grouped by status.'
      type: statusGrouped
      metricId: jira.open_issues
```

Updated to match the config from the plugin

Author: done

----
+
where:
+
`scorecard.aggregations.id`:: Specifies a unique identifier for the aggregated KPI.
`scorecard.aggregations.name`:: Specifies the name displayed on the homepage card.
`scorecard.aggregations.description`:: Specifies A brief explanation of what the metric represents.
`scorecard.aggregations.type`:: Specifies the aggregation function. Supported values include `sum` and `average`.
`scorecard.aggregations.metricId`:: Specifies the ID of the source metric from individual scorecards.

> `scorecard.aggregations.selector`:: Specifies the criteria used to group entities, typically defined by `owner`.

Reviewer 2: ```suggestion
`scorecard.aggregationKPIs`:: Map of KPI keys to KPI definitions. Each key is the aggregation id (for example `openIssuesKpi`).
`title`:: Title shown for this KPI.
`description`:: Short description of the KPI.
`type`:: Aggregation strategy: use `statusGrouped` for counts per threshold status, or `average` for a weighted portfolio score.
`metricId`:: Metric provider id (for example `jira.open_issues`, `github.open_prs`).
```

Updated to match the config from the plugin

Author: done

> . Save the configuration and restart your Red Hat Developer Hub instance. [cite: 5360, 3931]

Reviewer 2: ```suggestion
. Save the configuration and restart your {product} instance. [cite: 5360, 3931]
```

Another consistency fix

Author: done
